bindings/experimental/distrdf/python/DistRDF/HeadNode.py

```diff
@@ -0,0 +1,377 @@
+
+            friend_file_names (list): A list containing the file names
+                corresponding to a given treename in friend_names. Each
+                treename can correspond to multiple file names.
+        """
+        self.friend_names = friend_names
+        self.friend_file_names = friend_file_names
+
+    def __bool__(self):
+        """
+        Define the behaviour of FriendInfo instance when boolean evaluated.
+        Both lists have to be non-empty in order to return True.
+
+        Returns:
+            bool: True if both lists are non-empty, False otherwise.
+        """
+        return bool(self.friend_names) and bool(self.friend_file_names)
+
+    def __nonzero__(self):
+        """
+        Python 2 dunder method for __bool__. Kept for compatibility.
+        """
+        return self.__bool__()
+
+
+class Factory(object):
+    """
+    A factory for different kinds of root nodes of the RDataFrame computation
+    graph, depending on the arguments to the RDataFrame constructor. Currently
+    can create a TreeHeadNode or an EntriesHeadNode.
+    """
+
+    @staticmethod
+    def get_headnode(*args):
+        """
+        Parses the arguments and compares them against the possible RDataFrame
+        constructors
+        """
+
+        # Early check that arguments are accepted by RDataFrame
+        ROOT.RDataFrame(*args)
+
+        firstarg = args[0]
+        if isinstance(firstarg, int):
+            # RDataFrame(ULong64_t numEntries)
+            return EntriesHeadNode(firstarg)
+        elif isinstance(firstarg, (ROOT.TTree, str)):
+            # RDataFrame(std::string_view treeName, filenameglob, defaultBranches = {})
+            # RDataFrame(std::string_view treename, filenames, defaultBranches = {})
+            # RDataFrame(std::string_view treeName, dirPtr, defaultBranches = {})
+            # RDataFrame(TTree &tree, const ColumnNames_t &defaultBranches = {})
+            return TreeHeadNode(*args)
+        else:
+            raise RuntimeError(
+                ("First argument {} of type {} is not recognised as a supported "
+                 "argument for distributed RDataFrame. Currently only TTree/Tchain "
+                 "based datasets or datasets created from a number of entries "
+                 "can be processed distributedly.").format(firstarg, type(firstarg)))
+
+
+class EntriesHeadNode(Node.Node):
+    """
+    The head node of a computation graph where the RDataFrame is created from
+    a number of entries.
+    """
+
+    def __init__(self, nentries):
+        """
+        Creates a new RDataFrame instance for the given arguments.
+
+        Args:
+            nentries (int): The number of entries this RDataFrame will process.
+        """
+        super(EntriesHeadNode, self).__init__(None, None)
+
+        self.nentries = nentries
+        # Set at creation of the dataframe, might be optimized by the backend
+        # in optimize_partitions
+        self.npartitions = 2
```

Comment:
2 has been in this code since a long time. The reasoning is that it's the minimum sensible number of partitions for a distributed execution. The case of 1 partition is kind of borderline. In principle it's easy to do in Spark, but depending on the implementation of other distributed backends it might be more tricky

Reply:
being discussed in the context of #8485